evm/src/trace/identifier/etherscan.rs

```diff
@@ -0,0 +1,91 @@
+use super::TraceIdentifier;
+use ethers::{
+    abi::{Abi, Address},
+    etherscan,
+    types::Chain,
+};
+use futures::stream::{self, StreamExt};
+use std::{borrow::Cow, path::PathBuf};
+use tokio::time::{sleep, Duration};
+use tracing::{trace, warn};
+
+/// A trace identifier that tries to identify addresses using Etherscan.
+pub struct EtherscanIdentifier {
+    /// The Etherscan client
+    client: Option<etherscan::Client>,
+}
+
+impl EtherscanIdentifier {
+    pub fn new(
+        chain: Option<impl Into<Chain>>,
+        etherscan_api_key: String,
+        cache_path: Option<PathBuf>,
+        ttl: Duration,
+    ) -> Self {
+        if let Some(cache_path) = &cache_path {
+            if let Err(err) = std::fs::create_dir_all(cache_path.join("sources")) {
+                warn!(target: "etherscanidentifier", "could not create etherscan cache dir: {:?}", err);
+            }
+        }
+
+        Self {
+            client: chain.and_then(|chain| {
+                etherscan::Client::new_cached(chain.into(), etherscan_api_key, cache_path, ttl).ok()
+            }),
+        }
+    }
+}
+
+impl TraceIdentifier for EtherscanIdentifier {
+    fn identify_addresses(
+        &self,
+        addresses: Vec<(&Address, Option<&Vec<u8>>)>,
+    ) -> Vec<(Address, Option<String>, Option<String>, Option<Cow<Abi>>)> {
```

Comment:
it would be worth making this a standalone type, clippy would also definitely complain here